src/hotspot/share/classfile/classLoader.cpp

```diff
@@ -1227,6 +1230,52 @@ char* ClassLoader::skip_uri_protocol(char* source) {
   return source;
 }
 
+static char decode_percent_encoded(const char *str, size_t& index) {
+    if (str[index] == '%'
+            && isxdigit(str[index + 1])
+            && isxdigit(str[index + 2])) {
```

Comment:
I think you need to bound check the index because the `uri` from `ClassLoaderExt::process_module_table `could just be a directory name, e.g. dir%, without the jar file name.

Reply:
`str` is still zero-terminated, so if `str[index] == '%'` is true, then `str[index + 1]` at least exists and might be `0`. If `isxdigit(str[index + 1])` is true then `str[index + 2]` at least exists and might be `0`, in which case `isxdigit()` is going to be false for it.

Reply:
So if `isxdigit(str[index + 1])` is false, it won't proceed to check `isxdigit(str[index + 2])`?
The bound check I'm thinking about is like the following:
```
if (index >= strlen(str) - 2) {
  return str[index];
}
```

Reply:
> So if `isxdigit(str[index + 1])` is false, it won't proceed to check `isxdigit(str[index + 2])`? The bound check I'm thinking about is like the following:

That is exactly correct.

> The bound check I'm thinking about is like the following:
> 
> ```
> if (index >= strlen(str) - 2) {
>   return str[index];
> }
> ```

This seems unnecessary as we already assume that the string is null-terminated in the caller, and, as I have demonstrated out earlier, null-terminated strings are quite safe to check character-by-character.